walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,180 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
+  private int meetingDuration;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // The total (common) free time we have to conduct the meeting.
+    LinkedList<TimeRange> freeTime = new LinkedList<TimeRange>();
+    freeTime.addAll(Arrays.asList(TimeRange.WHOLE_DAY));
+
+    meetingDuration = (int) request.getDuration();
+    LinkedList<String> optionalAttendees = new LinkedList<>(request.getOptionalAttendees());
+
+    // If duration of the meeting is greater than a whole day, then return an empty list.
+    if (meetingDuration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    // If there are no attendees then return the whole day (ie the free time we have till now).
+    if (request.getAttendees().isEmpty() && optionalAttendees.isEmpty()) {
+      return freeTime;
+    }
+
+    // To store all the events which the optional attendees are attending.
+    LinkedList<Event> optionalAttendeeEvents = new LinkedList<>();
+
+    /** Changes freeTime such that it now conatins all the free time slots when the meeting can
+     * happen, updating the events that optional attendees can attend.
+     */
+    freeSlotsWithMandatoryAttendees(
+        freeTime, events, optionalAttendeeEvents, optionalAttendees, request);
+
+    /** If free time is not empty and optional attendees are not empty
+     * then go ahead and look for time slots to have meeting with our optional attendees 
+     */
+    if (!freeTime.isEmpty() && !optionalAttendeeEvents.isEmpty()) {
+      freeTime =
+          freeSlotsWithOptionalAttendees(optionalAttendeeEvents, freeTime, optionalAttendees);
+    }
+
+    return freeTime;
+  }
+
+  // Loop through all the events given and modify freeTime List
+  private void freeSlotsWithMandatoryAttendees(
+      LinkedList<TimeRange> freeTime,
+      Collection<Event> events,
+      LinkedList<Event> optionalAttendeeEvents,
+      LinkedList<String> optionalAttendees,
+      MeetingRequest request) {
+
+    for (Event event : events) {
+      /* Check if the event has any optional attendees and
+       * save it in the optionalAtttendeeEvents list which we will deal with
+       * later in the freeSlotsWithOptionalAttendees method
+       */
+      if (!Collections.disjoint(optionalAttendees, event.getAttendees()))
+        optionalAttendeeEvents.add(event);
+      else findConflictAndResolve(freeTime, new LinkedList<>(request.getAttendees()), event);
```

Comment:
For every event you create a new copy of the attendees. This is unnecessary and just costs time.

Reply:
True that, I am now passing the mandatoryAttendees variable and not creating a new one each time. 
Thanks for pointing out.